src/github.com/matrix-org/go-neb/services/google/google.go

```diff
@@ -0,0 +1,221 @@
+	querySentence := strings.Join(args, " ")
+
+	searchResult, err := s.text2imgGoogle(querySentence)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var imgURL = searchResult.Link
+	if imgURL == "" {
+		return gomatrix.TextMessage{
+			MsgType: "m.text.notice",
+			Body:    "No image found!",
+		}, nil
+	}
+
+	// FIXME -- Sometimes upload fails with a cryptic error - "msg=Upload request failed code=400 "
+	resUpload, err := client.UploadLink(imgURL)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to upload Google image to matrix: %s", err.Error())
+	}
+
+	img := searchResult.Image
+	return gomatrix.ImageMessage{
+		MsgType: "m.image",
+		Body:    querySentence,
+		URL:     resUpload.ContentURI,
+		Info: gomatrix.ImageInfo{
+			Height:   uint(math.Floor(img.Height)),
+			Width:    uint(math.Floor(img.Width)),
+			Mimetype: searchResult.Mime,
+		},
+	}, nil
+}
+
+// text2imgGoogle returns info about an image
+func (s *Service) text2imgGoogle(query string) (*googleSearchResult, error) {
+	log.Info("Searching Google for an image of a ", query)
+
+	u, err := url.Parse("https://www.googleapis.com/customsearch/v1")
+	if err != nil {
+		return nil, err
+	}
+
+	q := u.Query()
+	q.Set("q", query)            // String to search for
+	q.Set("num", "1")            // Just return 1 image result
+	q.Set("start", "1")          // No search result offset
+	q.Set("imgSize", "medium")   // Just search for medium size images
+	q.Set("searchType", "image") // Search for images
+	// q.set("fileType, "")                             // Any file format
+
+	var key = s.APIKey
+	if key == "" {
+		key = "AIzaSyA4FD39m9pN-hiYf2NRU9x9cOv5tekRDvM" // FIXME -- Should be instantiated from service config
+	}
+	q.Set("key", key)                                // Set the API key for the request
+	q.Set("cx", "003141582324323361145:f5zyrk9_8_m") // Set the custom search engine ID
+
+	u.RawQuery = q.Encode()
+	// log.Info("Request URL: ", u)
+
+	res, err := http.Get(u.String())
+	if res != nil {
+		defer res.Body.Close()
+	}
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode > 200 {
+		return nil, fmt.Errorf("Request error: %d, %s", res.StatusCode, response2String(res))
+	}
+	var searchResults googleSearchResults
+
+	// log.Info(response2String(res))
+	if err := json.NewDecoder(res.Body).Decode(&searchResults); err != nil || len(searchResults.Items) < 1 {
+		// Google return a JSON object which has { items: [] } if there are 0 results.
+		// This fails to be deserialised by Go.
+
+		// TODO -- Find out how to just return an error string (with no formatting)
+		// return nil, errors.New("No images found")
+		// return nil, fmt.Errorf("No results - %s", err)
+		return nil, fmt.Errorf("No images found%s", "")
+	}
+
+	// Return only the first search result
+	return &searchResults.Items[0], nil
+}
+
+// response2String returns a string representation of an HTTP response body
+func response2String(res *http.Response) (responseText string) {
```

Comment:
Ahh the joys of named-return values. You've given the return arg a name: `responseText`. This is not just for documentation purposes, it does Magic which can be convenient at best, and confusing at worst.

If you use a named-return values, when the function executes it will first zero-initialise the named return args and pull it into the function scope. That is to say, these 2 functions are equivalent:

```go
// Normal function
func response2String(res *http.Response) string {
    return "a response string"
}

// Named-return function
func response2String(res *http.Response) (responseText string) {
    // note the missing :
    responseText = "a response string"
    return
}
```

The named-return function also seems weird because it just has a `return` on its own at the bottom there, which is known as a "naked" return. Go knows at compile time **what** will be returned from the function, `responseText`, so it lets you omit that from the `return`. It doesn't know **when** it should return though, so it still needs you to use the `return` keyword. You've done a mixture of both, specifying a named-return but returning something else explicitly. This works but is confusing: stick to one form or the other.

So why use named-return functions? For convenience when you would otherwise need to initialise the return args, and for documentation purposes. See https://tour.golang.org/basics/7 for more information.